lib/targets/hip/device.cpp

```diff
@@ -27,136 +23,151 @@ namespace quda
+      for (int i = 0; i < get_device_count(); i++) {
+        CHECK_HIP_ERROR(hipGetDeviceProperties(&deviceProp, i));
+        if (getVerbosity() >= QUDA_SUMMARIZE) { printfQuda("Found device %d: %s\n", i, deviceProp.name); }
       }
 
       if (getVerbosity() >= QUDA_SUMMARIZE) { printfQuda("Using device %d: %s\n", dev, deviceProp.name); }
-#ifndef USE_QDPJIT
-      cudaSetDevice(dev);
-      checkCudaErrorNoSync(); // "NoSync" for correctness in HOST_DEBUG mode
-#endif
 
-#ifdef NUMA_NVML
-      char *enable_numa_env = getenv("QUDA_ENABLE_NUMA");
-      if (enable_numa_env && strcmp(enable_numa_env, "0") == 0) {
-        if (getVerbosity() > QUDA_SILENT) printfQuda("Disabling numa_affinity\n");
-      } else {
-        setNumaAffinityNVML(dev);
-      }
-#endif
+      CHECK_HIP_ERROR(hipSetDevice(dev));
+
+      CHECK_HIP_ERROR(hipDeviceSetCacheConfig(hipFuncCachePreferL1));
+      // hipDeviceSetSharedMemConfig(hipSharedMemBankSizeEightByte);
+      CHECK_HIP_ERROR(hipGetDeviceProperties(&deviceProp, dev));
+    }
 
-      cudaDeviceSetCacheConfig(cudaFuncCachePreferL1);
-      // cudaDeviceSetSharedMemConfig(cudaSharedMemBankSizeEightByte);
-      // cudaGetDeviceProperties(&deviceProp, dev);
+    int get_device_count()
+    {
+      static int device_count = 0;
+      if (device_count == 0) {
+        CHECK_HIP_ERROR(hipGetDeviceCount(&device_count));
+        if (device_count == 0) errorQuda("No HIP devices found");
+      }
+      return device_count;
     }
 
     void create_context()
     {
-      streams = new qudaStream_t[Nstream];
+      streams = new hipStream_t[Nstream];
 
       int greatestPriority;
       int leastPriority;
-      cudaDeviceGetStreamPriorityRange(&leastPriority, &greatestPriority);
+      CHECK_HIP_ERROR(hipDeviceGetStreamPriorityRange(&leastPriority, &greatestPriority));
       for (int i = 0; i < Nstream - 1; i++) {
-        cudaStreamCreateWithPriority(&streams[i], cudaStreamDefault, greatestPriority);
+        CHECK_HIP_ERROR(hipStreamCreateWithPriority(&streams[i], hipStreamDefault, greatestPriority));
       }
-      cudaStreamCreateWithPriority(&streams[Nstream - 1], cudaStreamDefault, leastPriority);
-
-      checkCudaError();
+      CHECK_HIP_ERROR(hipStreamCreateWithPriority(&streams[Nstream - 1], hipStreamDefault, leastPriority));
     }
 
     void destroy()
     {
       if (streams) {
-        for (int i = 0; i < Nstream; i++) cudaStreamDestroy(streams[i]);
+        for (int i = 0; i < Nstream; i++) CHECK_HIP_ERROR(hipStreamDestroy(streams[i]));
         delete[] streams;
         streams = nullptr;
       }
 
       char *device_reset_env = getenv("QUDA_DEVICE_RESET");
       if (device_reset_env && strcmp(device_reset_env, "1") == 0) {
-        // end this CUDA context
-        cudaDeviceReset();
+        // end this HIP context
+        CHECK_HIP_ERROR(hipDeviceReset());
       }
     }
 
+    hipStream_t get_cuda_stream(const qudaStream_t &stream) { return streams[stream.idx]; }
```

Comment:
You can nuke `get_cuda_stream`.  It was a temporary hack that I've removed.

Reply:
Nuked in a320a8